components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java

```diff
@@ -522,4 +540,390 @@ public RolesGetResponse listRolesWithPost(SearchRequest searchRequest)
+                    throw new BadRequestException(e.getMessage());
+                }
+                throw new CharonException(
+                        String.format("Error occurred while updating permissions for role: %s", roleId), e);
+            }
+        }
+    }
+
+    private void prepareAddedRemovedGroupLists(Set<String> addedGroupsIds, Set<String> removedGroupsIds,
+                                               Set<String> replacedGroupsIds, PatchOperation groupOperation,
+                                               Map<String, String> groupObject, List<GroupBasicInfo> groupListOfRole) {
+
+        switch (groupOperation.getOperation()) {
+            case (SCIMConstants.OperationalConstants.ADD):
+                removedGroupsIds.remove(groupObject.get(SCIMConstants.CommonSchemaConstants.VALUE));
+                if (!isGroupExist(groupObject.get(SCIMConstants.CommonSchemaConstants.VALUE), groupListOfRole)) {
+                    addedGroupsIds.add(groupObject.get(SCIMConstants.CommonSchemaConstants.VALUE));
+                }
+                break;
+            case (SCIMConstants.OperationalConstants.REMOVE):
+                addedGroupsIds.remove(groupObject.get(SCIMConstants.CommonSchemaConstants.VALUE));
+                removedGroupsIds.add(groupObject.get(SCIMConstants.CommonSchemaConstants.VALUE));
+                break;
+            case (SCIMConstants.OperationalConstants.REPLACE):
+                replacedGroupsIds.add(groupObject.get(SCIMConstants.CommonSchemaConstants.VALUE));
+                break;
+        }
+    }
+
+    private void prepareAddedRemovedUserLists(Set<String> addedMembers, Set<String> removedMembers,
+                                              Set<Object> newlyAddedMemberIds, PatchOperation memberOperation,
+                                              Map<String, String> memberObject, String currentRoleName)
+            throws UserStoreException, BadRequestException {
+
+        AbstractUserStoreManager userStoreManager =
+                (AbstractUserStoreManager) PrivilegedCarbonContext.getThreadLocalCarbonContext().getUserRealm()
+                        .getUserStoreManager();
+        if (StringUtils.isEmpty(memberObject.get(SCIMConstants.RoleSchemaConstants.DISPLAY))) {
+            List<org.wso2.carbon.user.core.common.User> userListWithID =
+                    userStoreManager.getUserListWithID(SCIMConstants.CommonSchemaConstants.ID_URI,
+                            memberObject.get(SCIMConstants.CommonSchemaConstants.VALUE), null);
+            if (isNotEmpty(userListWithID)) {
+                memberObject.put(SCIMConstants.RoleSchemaConstants.DISPLAY, userListWithID.get(0).getUsername());
+                memberOperation.setValues(memberObject);
+            }
+        }
+
+        if (memberObject.get(SCIMConstants.RoleSchemaConstants.DISPLAY) == null) {
+            throw new BadRequestException("User is not exist in any role.");
```

Comment:
"User is not exist in any role." is this correct? Shouldn't is be some thing like user is not exists in the userstore ?

Reply:
Fixed